README.md
# CETRIS

A Tetris game developed in C.
You can play in solo or you can watch Jarvis (The name of our AI) play alone.
This project was developed at ENSEIRB-MATMECA (Bordeaux, FRANCE) in a
course of initiation in C.

## Getting Started

This is a C project and we use Make to compile it.
Please follow the nexts steps.

### Prerequisites

+ Make to build the app
+ Curses library
+ Color suported terminal

### Installing

Download the project and extract it.
Go in the main folder with a terminal and type 'make all' to compile the app
Launch the game by typing './Cetris'

### Play !

Differents modes are available in the main menu :
+ *LOGIN* : Edit your login, it will appear in the scoreboard
+ PLAY : Normal mode
+ B MODE : Begin the game with an half filled board
+ NO WALL : Play without right and left walls
+ QUIT : Quit the game :(

In the 3 gameplay, you can play in solo or give the commands to Jarvis the AI
He plays very well to the game.
You can increase his speed by pressing 'H', take care of your processor !
In our last test, he placed 10 millions of tetri in 1 day and 6 hours. (Speed avg: 90tetri/sec)

## Good to know

+ Board has a fixed area of 255 pixels
+ Config file are saved near the binary
- Multi-instance of Cetris is possible
+ 2500 lines of code
+ Jarvis is not a real Artificial Inteligence
- No deep learning, he just tests all possibilities
- No optimisation when a branch seems to be a bad idea
+ Mesure working time and sleep '1ms - work_time' to get 1ms period

## Make's options

+ 'make all'
- Standard version of Cetris
+ 'make debug'
- No change from standard
+ 'make spymode'
- Enable AI_SPY_MODE to look how AI process tests

## Build options

+ NB_TETRI_COMMING : Number of tetri in the vision (Default 3)
- Also the number of tetri which are compute by Jarvis
+ POS_GUIDE_LVL_MAX : Level where guide lines are removed (Default 2)
+ DROP_SPEED : Speed of tetri when they are drop. Faster : 1 (Default 8)
+ AI_SPY_MODE : Force render when AI works (Default Not defined)
+ LOG_USE_COLOR : Use color in log file
+ DEBUG : Enable debug mode


## Built With

* [NCurses](https://www.gnu.org/software/ncurses/) - The graphic library used

## Authors

* **David DEVANT** - *Initial work* - [Devdevdavid](https://github.com/Devdevdavid)
* **Aurélien TROMPAT** - *Initial work* - [DjNounours](https://github.com/DjNounours)

## Improvements

If you want to add functions to the game you are welcome !






doc/changelog
[Revision 43] D.Devant (17/09/18)
DONE:
  > Added -gl -gs and -e for goal system
  
[Revision 43] D.Devant (25/01/17)
DONE:
  > Added ReadMe.txt
  > Updated PATH_MAX_LEN to 201
  > Bug fixe in cetris_path()
  > Updated AI_SPY_MODE
  > Added 'Q' to quit start menu
  > Improved time system
  > Added optimization option for make

[Revision 42] D.Devant (23/01/17)
DONE:
  > Minor display change
  > Added HighSpeed mode with 'H' key when AI is enable
  > Added empty login checkup
  > Bug fix in save_stats()
  > Bug fix in format_score()
  > Changed >= to > in update_topten()
TODO:
  > A good time system (Begun with time_spent in main)

[Revision 40 & 41] A.Trompat (20/01/17)
DONE:
  > Validate login modification on Windows OS and linux
  > Add tips on main menu
  > Modify Bumpiness multiplying factor to improve AI
  > Enable multi user for "cetris.conf"
  > Add speed counter tetri/sec
  > Fix bug on stats display
  
[Revision 39] D.Devant (18/01/17)
DONE:
  > Added Top 10 list updater
  > Remove old stats system
  > Updated display
TODO:
  > Try login modification on Windows OS
  > Fix speed counter without thread
  > Add tips on main menu
  > Enable multi user for "cetris.conf"

[Revision 38] D.Devant (17/01/17)
DONE:
  > Added login selection
  > Added login display
  > Added Top 10 in conf
  > Display Top 10
  > Defined AI's name as Jarvis
TODO:
  > Update order of Top 10 list


[Revision 37] D.Devant (13/01/17)
DONE:
  > Improve code visibility
  > Removing High Speed version

[Revision 36] D.Devant (13/01/17)
DONE:
  > New AI calculator
    - Remove completes lines
  > High speed version
    - Decreased FPS
    - Speed counter tetri/sec
    - Externalized render() with thread
  > Updated sevral function changing game_t to board_t

[Revision 33] D.Devant (11/01/17)
DONE:
  > Fixed bug with ESC key which introduces delay after press
  > Improve code visibility

[Revision 30 & 32] A.Trompat (10/01/17)
DONE:
	> Add counter of tetri placed and max tetri placed
	> Display all next tetri
	> Add possibility to rotate tetri when it is at the right border
	> Fix format_score
	> Fix calculation of number of positions of each tetri by ai
	
[Revision 31] D.Devant (10/01/17)
DONE:
  > New "Cleaner bonus" when player clear board (+3000pts)
  > Stop start menu blinking on ssh

[Revision 29] D.Devant (10/01/17)calcultating of number of positions of each tetri
DONE:
  > Change file path systeme

[Revision 28] A.Trompat (10/01/17)
	DONE:
	> uint32 for score and max score (uint16_t before)
	> Add log library

[Revision 27] D.Devant (07/01/17)
DONE:
  > Added new format for time display
  > Added log library
    - Added log messages pretty much everywhere
    - Logs are send to stderr and cetris.log
  > Fixed AI issues
    - Defined branch_t and node_t
      - A branch is an array of nodes
      - A branch have stats and a score
      - Just two things interrest us: max score and the first node of the best branch
  > Changed "ia" to "ai" everywhere !
  > Disabled print_debug(): not useful anymore
TODO:
  > Find a way to save files not at user position but at the binary position
  > SegFault on enseirb's server when quit
  > Show the vision number

[Revision 26] A.Trompat (07/01/17)
	> Anticipation de n piece par l'IA. Non fonctionnel
	bug 1 : l'ia fait les calculs mais ne positionne pas la piece au bonne endroit
	bug 2 : quand l'ia fait ses calculs et que les pieces sorte du board(en hauteur) le jeu ce bloc obbligé de faire CTRL+C
	
[Revision 23] D.Devant (20/12/17)
DONE :
  > Random choice of incomming_tetri
  > Init of incomming_tetri
  > Key to return menu
  > Add random_fill_board() for mode B
  > Modif of display
TODO :
  > Finish the expand of vision for AI

[Revision 23] D.Devant (20/12/17)
DONE :
  > Fix teleportation bug
  > Change name of tetri in used
  > Adding key to toggle AI player
  > Change time base of timeouts
  > Call callback function
TODO :
  > Random choice of incomming_tetri

[Revision 22] D.Devant (17/12/17)
DONE :
  > Change drop systeme to add animation with drop_speed
  > Add debug function: print branch result into file
  > Fix rotation by AI

[Revision 21] D.Devant && A.Trompat (13/12/17)
DONE :
  > Score calculate per branch with constant
  > Adding x, y and rotate to branch
  > Adding name for tetri
  > Adding improve_speed()
  > Modification for tab of branch
  > Adding erase_tetri()

[Revision 15] D.Devant (09/12/17)
DONE :
  > Come back to the elder file hiearchy
  > Delete of AI functions not needed anymore
  > Write of checking functions of AI
    - Count holes           VERIFIED
    - Count complete_lines  VERIFIED
    - Fill col_height tab   VERIFIED
    - Count aggredate       VERIFIED
    - Count bumpiness       VERIFIED
  > Create new function to refresh these variables
  > Create "#define DEBUG 1" with debug functions
    - Show col height
    - Show branch variables
TODO :
  > Change tetri in comming queue
  > Create tree with branch table




include/cetris.h
//
//  cetris.h
//  Cetris
//
//  Created by David Devant on 13/10/2017.
//  Copyright © 2017 David Devant. All rights reserved.
//

#ifndef CETRIS_H
#define CETRIS_H

#define VERSION               "v1.7.0 - October 2019 - Marvel Version"
#define LEVEL_MAX             10
#define MIN_BOARD_ROW         6
#define MIN_BOARD_COLUMN      4
#define NB_TETRI_COMMING      2
#define DROP_SPEED            8
#define POS_GUIDE_LVL_MAX     2

#define TERM_WIDTH            80
#define TERM_HEIGHT           24
#ifndef BOARD_WIDTH
	#define BOARD_WIDTH       8
#endif
#ifndef BOARD_HEIGHT
	#define BOARD_HEIGHT      20
#endif
#define BOARD_TABLE_SIZE      (BOARD_WIDTH * BOARD_HEIGHT)
#define PATH_MAX_LEN          201
#define ALT_BACKSPACE         127
#define ALT_ENTER             10
#define LOGIN_LEN             12
#define NB_MENU_ENTRIES       5
#define NB_TETRI              7

#define COLI_SUPER            0x01
#define COLI_UNDER            0x02
#define COLI_HORIZ_BORDER     0x04
#define COLI_VERTI_BORDER     0x08
#define COLI_BORDER           (COLI_HORIZ_BORDER | COLI_VERTI_BORDER)
#define COLI_ALL              0x0F

#define TITLE_LOGIN           0
#define TITLE_CLASSIC         1
#define TITLE_B_MODE          2
#define TITLE_NO_WALL         3
#define TITLE_QUIT            4

#define SCORE_CLEANER         3000
#define SCORE_PER_LEVEL       400
#define SCORE_COMPLETE_ROW    100
#define SCORE_WRITE_TETRI     5
#define SCORE_DROP            1

#define AI_SCORE_COMPL        +0.760666
#define AI_SCORE_AGGRE        -0.510066
#define AI_SCORE_HOLES        -0.356630
#define AI_SCORE_BUMPI        -0.484483

#define CONF_FILE_NAME        "cetris.conf"
#define LOG_FILE_NAME         "cetris.log"

// In seconds
#define BONUS_DELAY_MULTIPLY  10

#define NB_TIMEOUT            3
#define ID_TIMEOUT_MULTIPLY   0
#define ID_TIMEOUT_CLEANER    1
#define ID_TIMEOUT_SPEED      2

#define FREE                  0
#define DISABLED              0
#define ON_BOARD              0
#define MOVE_LATER            0
#define MOVE_NOW              1
#define LAST_TETRI            NB_TETRI_COMMING - 1

// Data Pipe
#define MAX_PIPE_NAME_LEN 	  32

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <unistd.h>
#include <curses.h>
#include <stdbool.h>
#include <string.h>
#include <time.h>
#include <pthread.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <errno.h>
#include "log.h"

struct block_t {
  char color;                   // Color is independant of shape
  uint8_t shape;                // Use 6 bit only
  char w;
  char h;
  char name;
};

struct tetri_t {
  uint8_t shape[6];
  uint8_t x;
  uint8_t y;
  uint8_t w;
  uint8_t h;
  uint8_t div_clk;              // For falling speed
  bool is_dropping;
  char name;
};

struct timer_t {
  time_t timeout[NB_TIMEOUT];   // Scheduled event
  time_t start;                 // Time from last pause
  time_t chrono;                // Game time
  bool is_running;
};

struct topten_t {
  char     login[LOGIN_LEN];
  uint32_t score;
  uint32_t tetri;
};

struct stats_t {
  struct topten_t topten[10];
  uint8_t  top_position;        // User's position in TOP 10
  uint32_t score;
  uint32_t tetri_placed;
  uint32_t tetri_placed_old;    // Old value to compute placement speed
  uint16_t speed;
  uint8_t  level;
  uint8_t  multiply;
  bool     cleaner_msg_enable; // Flag for cleanner bonus message
  char     conf_path[PATH_MAX_LEN];
};

struct board_t {
  uint8_t w;
  uint8_t h;
  bool no_wall_mode;
  uint8_t table[BOARD_TABLE_SIZE];
};

struct display_t {
  WINDOW *start_win;
  WINDOW *board;
  WINDOW *next_tetri;
  WINDOW *stats;
  WINDOW *hotkey;
  WINDOW *topten;
  WINDOW *status;
  WINDOW *goal_msg;
  uint8_t is_goal_msg_visible;
  uint8_t select;               // Menu selection
  char login[LOGIN_LEN];
};

struct node_t {
  uint8_t   x;
  uint8_t   y;
  uint8_t   rotate;
  uint8_t   complete_line;
};

struct branch_t {
  struct node_t nodes[NB_TETRI_COMMING];
  uint8_t   holes;              // number of holes
  uint8_t   complete_lines;     // 4 max
  uint16_t  aggregate_height;   // Sum of all column height
  uint16_t  bumpiness;          // Sum of height differences
  uint8_t   col_height[20];
  double    score;
};

struct ai_brain_t {
  struct branch_t branch;
  struct node_t best_first_node;
  struct board_t board[NB_TETRI_COMMING];
  double best_branch_score;
  uint16_t cur_tetri;
  uint16_t nb_passage;
  bool enable;
  bool high_speed;
};

struct goal_t {
	uint8_t event; /* 0: Goal disabled, > 0: Event ID */
	uint32_t score;
	bool isScoreReached;
	uint8_t level;
	bool isLevelReached;
};

struct data_pipe_t {
	char pipeName[MAX_PIPE_NAME_LEN];
	int fifoFd;
	uint8_t tableCpy[BOARD_TABLE_SIZE];				/* Copy of board->table with the tetri in it */
};

struct game_t {
  struct stats_t stats;
  struct timer_t timer;
  struct tetri_t tetri_comming[NB_TETRI_COMMING];
  struct block_t blocks[NB_TETRI];
  struct board_t board;
  struct display_t display;
  struct ai_brain_t ai_brain;
  struct goal_t goal;
  struct data_pipe_t data_pipe;
  char cetris_path[PATH_MAX_LEN];
  bool pause;
  bool in_game;
  bool ended;
  bool quit;
  uint64_t cur_time;
};

/* ====================== *
 *         MAIN           *
 * ====================== */
void game_sleep(void);
void update(struct game_t *);
void place_tetri(struct game_t *);
void render(struct game_t *);
void check_goal(struct game_t *);

/* ====================== *
 *      CONFIG_FILE       *
 * ====================== */
void init_file(struct game_t *);
void init_log(struct game_t *);
void save_stats(struct game_t *);
void read_file_stats(struct stats_t *);
void write_file_stats(struct stats_t *);
void get_cetris_path(struct game_t *, const char * first_arg);
void get_label(char buffer[], char label[]);
char * get_string(char buffer[]);
uint32_t get_value(char buffer[]);
uint16_t get_file_len(char path[]);

/* ====================== *
 *         STATS          *
 * ====================== */
void init_stats(struct stats_t *);
void set_score(struct game_t *, uint32_t new_score);
void add_tetri_placed(struct stats_t *);
void add_score(struct game_t *, uint16_t points);
void level_up(struct game_t *);
void set_cleaner_bonus(struct game_t *);
void set_multiply(struct game_t *, uint8_t multiply, uint16_t duration);
void update_topten(struct game_t *);
void bubble_sorting(struct topten_t *);
void swap_topten(struct topten_t *t1, struct topten_t *t2);
void update_speed_counter(struct game_t *);

/* ====================== *
 *         TIMER          *
 * ====================== */
void init_timer(struct timer_t *);
void start_timer(struct timer_t *);
void stop_timer(struct timer_t *);
struct tm* get_timer(struct timer_t *);
void set_timeout(struct timer_t *, uint8_t id, uint16_t duration);
void timeout_callback(struct game_t *);

/* ====================== *
 *     BOARD ANALYSE      *
 * ====================== */
void init_block(struct block_t *);
void init_board(struct board_t *);
void erase_line(struct board_t *, uint8_t y);
void check_complete_lines(struct game_t *);
void write_tetri(struct board_t *, struct tetri_t *);
void erase_tetri(struct board_t *, struct tetri_t *);
void add_row_board(struct game_t *, uint8_t row);
void remove_row_board(struct game_t *, uint8_t row);
void add_column_board(struct game_t *, uint8_t column);
void remove_column_board(struct game_t *, uint8_t column);
void random_fill_board(struct game_t *);
bool check_end_game(struct game_t *);
void clear_board(struct board_t *);
bool is_line_free(struct board_t *, uint8_t y);
bool is_line_complete(struct board_t *, uint8_t y);
void swap_line(struct board_t *, uint8_t y1, uint8_t y2);
void copy_tab(uint8_t *tab_dest, uint8_t *tab_src, uint16_t length);
bool board_size_is_valid(uint8_t w, uint8_t h);


/* ====================== *
 *       TETRIMINO        *
 * ====================== */
void init_tetri(struct game_t *);
uint8_t rotate90_tetri(struct board_t *, struct tetri_t *);
void move_tetri_left(struct board_t *, struct tetri_t *);
void move_tetri_right(struct board_t *, struct tetri_t *);
void move_tetri_down(struct board_t *, struct tetri_t *);
bool test_fall_speed(struct game_t *);
bool test_drop_speed(struct game_t *);
bool check_tetri_colision(struct board_t *, struct tetri_t *,  uint8_t filter);
void drop_tetri(struct board_t *, struct tetri_t *);
void set_drop_flag_tetri(struct tetri_t *);
void dropping_tetri(struct game_t *);
void use_next_tetri(struct game_t*);

// Internal Functions
void swap_u8(uint8_t *a, uint8_t *b);
void transpose_shape(struct tetri_t *);
void reverse_column(struct tetri_t *);
void cpy_block_to_tetri(struct tetri_t *, struct block_t *);
void choose_next_tetri(struct game_t*);

/* ====================== *
 *       USER_CMD         *
 * ====================== */
void init(struct game_t *);
void reset(struct game_t *);
void start(struct game_t *);
void stop(struct game_t *);
void quit(struct game_t *);
void return_menu(struct game_t *);
void check_user_input(struct game_t *);

// Internal Functions
void check_user_input_start(struct game_t *);
void check_user_input_game(struct game_t *);
void edit_login(struct game_t *, uint16_t key);
void process_choice(struct game_t *);
void check_login(struct display_t *);
void toggle_pause(struct game_t *);
void to_upper(char s[]);

/* ====================== *
 *        DISPLAY         *
 * ====================== */
bool init_display(struct game_t *);
void update_windows_position(struct game_t *);
void clear_display(struct display_t *);
void print_start_win(struct display_t *);
void print_play_menu(struct display_t *);
void print_game_win(struct game_t *);

// Window print
void print_board(struct game_t *);
void print_tetri(struct game_t *);
void print_position_guide(struct game_t *);
void print_next_tetri(struct game_t *);
void print_stats(struct game_t *);
void print_hotkey(struct display_t *);
void print_topten(struct game_t *);
void print_status(struct game_t *);
void print_goal_msg(struct game_t *);
void apply_prints(struct game_t *);

// Internal Functions
void init_display_color(void);
void print_pixel(WINDOW *, uint8_t color, uint8_t x, uint8_t y);
void print_box(WINDOW *, char title[]);
char * format_score(uint32_t score, char buffer[]);
char * format_time(struct tm timeinfo, char *buffer);
char * add_spaces_after_login(char src[], char dest[]);

/* ====================== *
 *       AI_PLAYER        *
 * ====================== */
void ai_init(struct ai_brain_t *);
void toggle_ai(struct game_t *game);
void toggle_high_speed(struct game_t *);
void ai_move(struct game_t *);

// Internals Functions
uint16_t ai_compute_node(struct game_t *);
void ai_get_holes(struct branch_t *, struct board_t *);
void ai_get_column_height(struct branch_t *, struct board_t *);
void ai_compute_score(struct branch_t *);
void improve_speed(struct tetri_t *, uint8_t *r_index);
uint8_t remove_complete_line(struct board_t *);
void apply_node(struct game_t *, struct node_t *);
void log_best_score(struct ai_brain_t *);
void log_ia_boards(struct ai_brain_t *);

/* ====================== *
 *       DATA_PIPE        *
 * ====================== */
int init_data_pipe(struct data_pipe_t *data_pipe);
int data_pipe_close(struct data_pipe_t *data_pipe);
int data_pipe_write(struct game_t *game);

#endif /* CETRIS_H */







